lightning/src/util/persist.rs

```diff
@@ -0,0 +1,81 @@
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Traits for handling persistence
+
+use core::ops::Deref;
+use bitcoin::hashes::hex::ToHex;
+use prelude::*;
+use io::{self};
+
+use crate::{chain::{keysinterface::{Sign, KeysInterface}, self, transaction::{OutPoint}, chaininterface::{BroadcasterInterface, FeeEstimator}, chainmonitor::{Persist, MonitorUpdateId}, channelmonitor::{ChannelMonitor, ChannelMonitorUpdate}}, ln::channelmanager::ChannelManager, routing::network_graph::NetworkGraph};
+
+use super::{logger::Logger, ser::Writeable};
+
+/// Trait that handles persisting a [`ChannelManager`] and [`NetworkGraph`] to disk.
+pub trait Persister<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> where
+M::Target: 'static + chain::Watch<Signer>,
+T::Target: 'static + BroadcasterInterface,
+K::Target: 'static + KeysInterface<Signer = Signer>,
+F::Target: 'static + FeeEstimator,
+L::Target: 'static + Logger
+{
+	/// Persist the given ['ChannelManager'] to disk, returning an error if persistence failed.
+	fn persist_manager(&self, channel_manager: &ChannelManager<Signer, M, T, K, F, L>) -> Result<(), io::Error>;
+
+	/// Persist the given [`NetworkGraph`] to disk, returning an error if persistence failed.
+	fn persist_graph(&self, network_graph: &NetworkGraph) -> Result<(), io::Error>;
+}
+
+/// Trait for a key-value store for persisting some writeable object at some key
+pub trait KVStorePersister {
+	/// Persist the given writeable using the provided key
+	fn persist<W: Writeable>(&self, key: String, object: &W) -> io::Result<()>;
+	/// Get the key for persisting a channel manager
+	fn get_channel_manager_key(&self) -> String { String::from("manager") }
+	/// Get the key for persisting a network graph
+	fn get_network_graph_key(&self) -> String { String::from("network_graph") }
+	/// Get the key for persisting a channel monitor from it's funding txo
+	fn get_channel_monitor_key(&self, funding_txo: &OutPoint) -> String {
+		format!("monitors/{}_{}", funding_txo.txid.to_hex(), funding_txo.index)
+	}
+}
+
+impl<A: KVStorePersister, Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> Persister<Signer, M, T, K, F, L> for A 
+where 
```

Comment:
nit: spurious whitespace at EOL here and a few other places. I'd generally suggest glancing with a local `git show` which should highlight such whitespace depending on your terminal settings.

Reply:
why not just use rustfmt for the project and run on pre-commit or whatever?

Reply:
Sadly rustfmt tends to generate garbage code formatting IMO, doubly so on our codebase. https://github.com/lightningdevkit/rust-lightning/issues/410#issuecomment-699570105